### PR TITLE
calculate metrics for p and k union

### PR DESCRIPTION
**Pull request by aws-zhenguo**

### CODE REVIEW

File: axlearn_inference.py [hunk 1]
```diff
@@ -283,7 +283,41 @@ def assert_top_k_allclose(x, y, k=64):
...
+    x_indices = np.argsort(x, axis=-1)[:,:,::-1]
+
+    top_x_flat = list()
+    top_y_flat = list()
+    for i, (x_seq, y_seq) in enumerate(zip(x, y)):
+        for j, (x_token, y_token) in enumerate(zip(x_seq, y_seq)):
+            top_indices = list()
+            total = 0.0
+            for idx in x_indices[i][j]:
+                total += x_token[idx]
```
Comment (HahTK) [inline axlearn_inference.py:299]: you also need top_p for y

top_p = top_p_y + top_p_x
ditto topk


Essentially you need 2 loops 
loop1 : do topP and K for X
loop2 : do topP and K for Y

topP/K = top_p/k_y + top_p/k_x

File: axlearn_inference.py [hunk 2]
```diff
@@ -267,7 +267,7 @@ def assert_top_p_allclose(x, y, p=0.99):
     print("max top p relative difference", np.max(rdiff))
     print("mean top p relative difference", np.mean(rdiff))
 
-    assert np.allclose(top_x, top_y, atol=1e-3, rtol=1e-2)
+    np.testing.assert_allclose(top_x, top_y, atol=1e-3, rtol=2e-2)
```
Comment (HahTK) [inline axlearn_inference.py:270]: we need to review the atol and rtol values and compare them against the smallest union of topP and topK on both x and y logits

File: axlearn_inference.py [hunk 3]
```diff
@@ -352,6 +386,7 @@ def pad_list(l, max_len, fill_value=tokenizer.pad_id):
     llama_probs = torch.softmax(output.logits, dim=-1).numpy()
```
Comment (HahTK) [inline axlearn_inference.py:386]: Would like a test where we get the softmax output from axlearn directly.
We should check that the logits used for that softmax input AND output is also in FP32.
It is ok to PR to an accuracy branch of axlearn to check the softmax input AND output dtype.

**Review by HahTK — COMMENTED** [review]
In addition to the specific items here. We should also have someone from infra do a full code review as well. I propose @kumpre

File: axlearn_inference.py (hunk 1 above)
Comment (aws-zhenguo) [reply axlearn_inference.py:299]: gotcha. PR created here https://github.com/aws-zhenguo/axlearn/pull/2

File: axlearn_inference.py (hunk 3 above)
Comment (aws-zhenguo) [reply axlearn_inference.py:386]: will work on it
